[float8nocompile] Simplified Float8Linear implementation which only supports dynamic tensorwise scaling

danielvegamyhre opened the pull request:
Summary:

This PR adds a simplified implementation of Float8Linear dubbed `Float8LinearNoCompile` which only supports dynamic tensorwise scaling. I've used TODOs to mark the places where I need to replace the torch based logic with custom triton kernels, in order to improve eager mode performance. Once those kernels have been implemented, I'll benchmark the performance and do some profiling to identify bottlenecks and find additional optimization opportunities. 

The purpose of starting with  this is to start with a simple implementation which works e2e for float8 training (as shown in the test plan section below), so that as I start replacing torch logic with new triton kernels, I can use the e2e training example as a basic test to validate it's still working.

Test plan:
- Validated this simplified implementation works e2e by running a simple example in `examples/example.py`:

```
[danvm@devgpu006.vll6 /data/users/danvm/ao/torchao/prototype/float8nocompile/examples (linear1)]$ python3 example.py 
calling convert_to_float8_nocompile_training
finished convert_to_float8_nocompile_training
step 0
step 1
step 2
step 3
step 4
step 5
step 6
step 7
step 8
step 9
```

vkuzo commented on torchao/prototype/float8nocompile/examples/example.py:
```diff
@@ -0,0 +1,33 @@
+convert_to_float8_nocompile_training(m)
+print("finished convert_to_float8_nocompile_training")
+
+for i in range(10):
```
this makes sense, I'd recommend adding a stronger numerical equivalency check (can be in a future PR):
1. create example input
2. create (a) reference model and (b) your model
3. feed the example input, run backwards
4. compare that model output, weight gradients, input gradients are equivalent across (a) and (b)

danielvegamyhre replied:
Makes sense, will do this in a follow up PR.

vkuzo commented on torchao/prototype/float8nocompile/float8nocompile_linear.py:
```diff
@@ -19,4 +36,149 @@ class Float8LinearNoCompile(torch.nn.Linear):
+        """
+
+        # Amax scales should always be kept as float32.
+        self.always_float32_buffers = set()
```
can delete, buffers are only needed for delayed scaling

danielvegamyhre replied:
ah yes I noticed buffers weren't needed for dynamic scaling but missed this one somehow - thanks!

vkuzo commented on torchao/prototype/float8nocompile/float8nocompile_linear.py:
```diff
@@ -19,4 +36,149 @@ class Float8LinearNoCompile(torch.nn.Linear):
+        emulate = config.emulate
+        super().__init__(*args, **kwargs)
+
+        # Defines the scaling behavior of input, weight, grad_output
```
can delete, this is only needed for supporting multiple types of scaling

vkuzo commented on torchao/prototype/float8nocompile/float8nocompile_linear.py:
```diff
@@ -19,4 +36,149 @@ class Float8LinearNoCompile(torch.nn.Linear):
+        self.scaling_type_grad_output = config.cast_config_grad_output.scaling_type
+
+        self.config = config
+        self.is_amax_initialized = not self.config.enable_amax_init
```
can delete, this is only needed for delayed scaling

vkuzo commented on torchao/prototype/float8nocompile/float8nocompile_linear.py:
```diff
@@ -19,4 +36,149 @@ class Float8LinearNoCompile(torch.nn.Linear):
+    def forward(self, input: torch.Tensor) -> torch.Tensor:
+        # TODO(danielvegamyhre): modify to support for FSDP once dependencies are implemented
+        output = self.forward_fp8_matmul(input)
+        if self.bias is not None:
```
feel free to not support bias at all to simplify, as our target model uses linear without bias

vkuzo commented on torchao/prototype/float8nocompile/float8nocompile_linear.py:
```diff
@@ -19,4 +36,149 @@ class Float8LinearNoCompile(torch.nn.Linear):
+
+    def get_weight_scale(self, weight: torch.Tensor) -> Optional[torch.Tensor]:
+        # TODO(danielvegamyhre): replace scale calculation with triton kernel
+        if tensor_already_casted_to_fp8(weight):
```
feel free to remove this to simplify

vkuzo commented on torchao/prototype/float8nocompile/float8nocompile_linear.py:
```diff
@@ -19,4 +36,149 @@ class Float8LinearNoCompile(torch.nn.Linear):
+    def from_float(
+        cls,
+        mod,
+        config: Optional[Float8LinearConfig] = None,
```
nit: feel free to only support default settings and not give an option to modify the config

vkuzo commented on torchao/prototype/float8nocompile/float8nocompile_scaling_utils.py:
```diff
@@ -0,0 +1,69 @@
+        axiswise_dim: if axiswise granularity is used, defines the dim to scale across
+    """
+    # TODO(danielvegamyhre): replace this torch implementation with custom triton kernel
+    if tensor_already_casted_to_fp8(hp_tensor):
```
nit: can remove to simplify

vkuzo commented on torchao/prototype/float8nocompile/float8nocompile_scaling_utils.py:
```diff
@@ -0,0 +1,69 @@
+    # TODO(danielvegamyhre): replace this torch implementation with custom triton kernel
+    if tensor_already_casted_to_fp8(hp_tensor):
+        return hp_tensor
+    scale = tensor_to_scale(
```
I'd recommend just inlining the right code here for just the default settings instead of calling into util functions, it will be easier to compare that to handwritten kernels

vkuzo commented on torchao/prototype/float8nocompile/float8nocompile_scaling_utils.py:
```diff
@@ -0,0 +1,69 @@
+        scaling_granularity,
+        axiswise_dim,
+    )
+    return hp_tensor_and_scale_to_float8(
```
I'd recommend just inlining the right code here for just the default settings instead of calling into util functions, it will be easier to compare that to handwritten kernels

vkuzo reviewed (approved):
lgtm, feel free to address changes in this PR or future PRs

danielvegamyhre commented:
> lgtm, feel free to address changes in this PR or future PRs

Sounds good, I addressed all your comments except the extension to the test script, which I'll do in a follow up PR - thanks!
